dvc/scm.py

```diff
@@ -47,6 +49,25 @@ def __init__(self, reason: str) -> None:
         super().__init__(f"{reason}\n{doc}")
 
 
+class GitMergeError(SCMError):
+    def __init__(self, msg: str, scm: Optional["Git"] = None) -> None:
+        if scm and self._is_shallow(scm):
+            url = format_link(
+                "https://dvc.org/doc/user-guide/troubleshooting#git-shallow"
+            )
+            msg = (
+                f"{msg}: Cannot merge in shallow Git repo. "
+                f"See {url} for more information."
```

Comment:
Just to clarify, is this a problem beyond `dvc exp` commands? Why are there internal Git merges again? Just FMI + maybe there's a better way to explain the issue (once I fully understand it). Thanks

Reply:
This does not affect anything besides `exp` commands. Internally we use merges when applying stash (merge) commits, which happens both on `exp apply` and `exp run`. When running experiments we generate the stash commit to store the experiment changes, and then apply (merge) them when the experiment is actually executed (either in the workspace or in the separate temp directory).

Reply:
OK so I'd make this more specific then. Maybe

```suggestion
                "https://dvc.org/doc/user-guide/troubleshooting#git-shallow"
            )
            msg = (
                f"{msg}:`dvc exp` doesn't work in shallow Git repos. "
                f"See {url} for more information."
```